### PR TITLE
CurriculumWrapper

### PR DESCRIPTION
**Pull request by emailweixu**

### CODE REVIEW

File: alf/environments/alf_wrappers.py [hunk 1]
```diff
@@ -465,3 +481,285 @@ def reward_spec(self):
     def time_step_spec(self):
         spec = self._env.time_step_spec()
         return spec._replace(reward=self.reward_spec())
+
+
+class MultitaskWrapper(AlfEnvironment):
+    """Multitask environment based a list of environments.
```
Comment (hnyu) [inline alf/environments/alf_wrappers.py]: based -> based on

File: alf/environments/alf_wrappers.py [hunk 2]
```diff
@@ -465,3 +481,285 @@ def reward_spec(self):
...
+
+    'task_id' is used to specify which task to run for the current step.
+    """
+
+    def __init__(self, envs, task_names):
+        """
+        Args:
+            envs (list[AlfEnvironment]): a list of environments. Each one
+                represents a different task.
+        """
```
Comment (hnyu) [inline alf/environments/alf_wrappers.py:511]: assert len(envs) > 0

File: alf/environments/alf_wrappers.py [hunk 3]
```diff
@@ -465,3 +481,285 @@ def reward_spec(self):
...
+
+        {
+            'task_id': TensorSpec((), maximum=num_envs - 1, dtype='int64'),
+            'action': original_action_spec
+        }
+
+    'task_id' is used to specify which task to run for the current step.
+    """
+
+    def __init__(self, envs, task_names):
```
Comment (hnyu) [inline alf/environments/alf_wrappers.py]: task_names can be an optional argument. If it's None, then env names are used instead.

File: alf/environments/alf_wrappers.py [hunk 4]
```diff
@@ -465,3 +481,285 @@ def reward_spec(self):
...
+            ``progress_favor`` during the first ``num_tasks * warmup_period``
+            episodes
+        """
+        self._env = env
+        assert env.batched, "Only batched env is supported"
+        num_tasks = env.num_tasks
+        task_names = env.task_names
+        batch_size = env.batch_size
+        self._episode_rewards = torch.zeros(batch_size)
+        assert (
```
Comment (hnyu) [inline alf/environments/alf_wrappers.py:658]: Should just require env to be a MultitaskWrapper instead.

File: alf/environments/suite_gym.py [hunk 5]
```diff
@@ -50,6 +50,9 @@ def load(environment_name,
     Returns:
         An AlfEnvironment instance.
     """
+    if isinstance(environment_name, (list, tuple)):
+        return alf_wrappers.MultitaskWrapper.load(load, **locals())
```
Comment (hnyu) [inline alf/environments/suite_gym.py]: This is not a scalable way of adding MultitaskWrapper? Other suite's load function need to be changed too.

File: alf/environments/alf_wrappers_test.py [hunk 6]
```diff
@@ -170,5 +172,78 @@ def test_duration_applied_after_episode_terminates_early(self):
         self.assertTrue(last_time_step.is_last())
 
 
+class MultitaskWrapperTest(alf.test.TestCase):
+    def test_multitask_wrapper(self):
+        env = alf_wrappers.MultitaskWrapper.load(
+            suite_gym.load, ['CartPole-v1', 'CartPole-v1'])
```
Comment (hnyu) [inline alf/environments/alf_wrappers_test.py]: Should check task names to be different because they are used as keys for env_info

File: alf/environments/alf_wrappers.py [hunk 7]
```diff
@@ -465,3 +481,285 @@ def reward_spec(self):
...
+            self._past_scores[task_id] += self._past_score_update_rate * (
+                task_score - self._past_scores[task_id])
+
+        progresses = (self._current_scores - self._past_scores).relu()
+        max_progress = progresses.max()
+        progresses = progresses / (max_progress + 1e-30)
+        # Gradually increase scale from 0 to self._scale so that we tend to do
+        # random smapling of the environments initially
+        scale = self._scale * min(1, self._total_count / self._warmup_period)
+        self._task_probs = F.softmax(scale * progresses, dim=0)
```
Comment (hnyu) [inline alf/environments/alf_wrappers.py:719]: This task_probs calculation is kind of unintuitive. The progresses are basically ratios but treated as logits after scaling?

File: alf/environments/alf_wrappers.py [hunk 8]
```diff
@@ -465,3 +481,285 @@ def reward_spec(self):
...
+
+    def __init__(self,
+                 env,
+                 progress_favor=10.0,
+                 current_score_update_rate=1e-3,
+                 past_score_update_rate=5e-4,
+                 warmup_period=100):
+        """
+        env (AlfEnvironment): environment to be wrapped. It needs to be batched.
+        progress_favor (float): how much more likely to choose the environment with the
```
Comment (hnyu) [inline alf/environments/alf_wrappers.py:643]: This number is multiplied to all progress ratios. So this description is not accurate.

File: alf/environments/alf_wrappers.py [hunk 9]
```diff
@@ -465,3 +481,285 @@ def reward_spec(self):
...
+        time_step = self._env.reset()
+        info = copy.copy(time_step.env_info)
+        info.update(self._zero_curriculum_info)
+        return time_step._replace(
+            env_info=info, prev_action=time_step.prev_action['action'])
+
+    def _step(self, action):
+        time_step = self._env.step(
+            OrderedDict(task_id=self._current_task_ids, action=action))
+        task_ids = time_step.prev_action['task_id']
```
Comment (hnyu) [inline alf/environments/alf_wrappers.py]: This should just be self._current_task_ids

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/environments/alf_wrappers.py (hunk 1 above)
Comment (emailweixu) [reply alf/environments/alf_wrappers.py]: Fixed

File: alf/environments/alf_wrappers.py (hunk 3 above)
Comment (emailweixu) [reply alf/environments/alf_wrappers.py]: Indeed, the task names are environment names passed in by load().
AlfEnvironment does not have name. So we cannot get them from envs.

File: alf/environments/alf_wrappers.py (hunk 2 above)
Comment (emailweixu) [reply alf/environments/alf_wrappers.py:511]: added

File: alf/environments/alf_wrappers.py (hunk 4 above)
Comment (emailweixu) [reply alf/environments/alf_wrappers.py:658]: It might be possible that some environment provides native multi-task support.

File: alf/environments/alf_wrappers.py (hunk 8 above)
Comment (emailweixu) [reply alf/environments/alf_wrappers.py:643]: This is indeed correct. The probability of the fastest one being sampled will be  ``progress_favor`` times more than the one with a task of 0 or negative progress (if there is one).

File: alf/environments/alf_wrappers.py (hunk 7 above)
Comment (emailweixu) [reply alf/environments/alf_wrappers.py:719]: progresses are normalized to [0, 1]

File: alf/environments/alf_wrappers.py (hunk 9 above)
Comment (emailweixu) [reply alf/environments/alf_wrappers.py]: Changed

File: alf/environments/alf_wrappers_test.py (hunk 6 above)
Comment (emailweixu) [reply alf/environments/alf_wrappers_test.py]: added

File: alf/environments/suite_gym.py (hunk 5 above)
Comment (emailweixu) [reply alf/environments/suite_gym.py]: Fixed.

**Review by hnyu — APPROVED** [review]
